File: fp_core/src/test/src/Elements/Playable.cpp
#include <FpCoreTest.h>
#include <gtest/gtest.h>
#include <fruitpunch/Common/Observable.h>
#include <fruitpunch/Common/Observer.h>
#include <fruitpunch/Elements/Playable.h>
#include <boost/shared_ptr.hpp>
#include <fruitpunch/Common/Events.h>


using namespace fp_core;
using namespace boost;

// ---------------------------------------------------------------------------
// Stub Classes
// ---------------------------------------------------------------------------

class PlayableImpl : public Playable {
public:
  bool renderSingleCalled;
  int frameRendered;

  PlayableImpl() : Playable() {
    renderSingleCalled = false;
  }

  virtual void renderSingle(int frame) {
    renderSingleCalled = true;
    frameRendered = frame;
  }

  virtual int calculateDuration() {
    return 10;
  }
};

class DummyObserver : public Observer {
public:

	bool triggered;

	virtual void action() {
		triggered = true;
	}
};

// ---------------------------------------------------------------------------
// Test Class
// ---------------------------------------------------------------------------

class PlayableTest : public FpCoreTest {
protected:
  virtual void SetUp() {
	  initEnvironment();

  }
};

// ---------------------------------------------------------------------------
// Test Cases
// ---------------------------------------------------------------------------

/**
 * Tests methods that have simple test cases
 */
TEST_F(PlayableTest, SmallMethods) {
  PlayableImpl p;

  ASSERT_EQ(p.getCurrentFrame(), 0);

  p.start();
  ASSERT_TRUE(p.isPlaying());

  p.stop();
  ASSERT_FALSE(p.isPlaying());
}

TEST_F(PlayableTest, tick) {
  PlayableImpl p;

  p.tick();
  ASSERT_FALSE(p.renderSingleCalled);
  ASSERT_EQ(0, p.getCurrentFrame());

  p.start();
  p.tick();
  ASSERT_TRUE(p.renderSingleCalled);
  ASSERT_EQ(1, p.getCurrentFrame());

  p.setCurrentFrame(10);
  p.tick();
  ASSERT_EQ(p.getCurrentFrame(), 0);

  p.stop();
  ASSERT_FALSE(p.isPlaying());

  p.setCurrentFrame(10);
  p.start();
  p.setLoop(true);
  p.tick();
  ASSERT_EQ(p.getCurrentFrame(), 0);
  ASSERT_TRUE(p.isPlaying());
}

TEST_F(PlayableTest, getEndFrame) {
  PlayableImpl p;

  ASSERT_EQ(9, p.getEndFrame());

  p.setStartFrame(10);

  ASSERT_EQ(19, p.getEndFrame());
}

TEST_F(PlayableTest, frameEvents) {
	boost::shared_ptr<PlayableImpl> p(new PlayableImpl());
	boost::shared_ptr<DummyObserver> first(new DummyObserver());
	boost::shared_ptr<DummyObserver> last(new DummyObserver());
	first->triggered = false;
	last->triggered = false;


	p->attachObserver(EV_FIRST_FRAME, first);
	p->attachObserver(EV_LAST_FRAME, last);

	p->setStartFrame(10);
	p->start();

	p->setCurrentFrame(9);
	p->tick();
	ASSERT_FALSE(first->triggered);
	ASSERT_FALSE(last->triggered);

	p->setCurrentFrame(10);
	p->tick();
	ASSERT_TRUE(first->triggered);
	ASSERT_FALSE(last->triggered);
	first->triggered = false;

	p->setCurrentFrame(19);
	p->tick();
	ASSERT_FALSE(first->triggered);
	ASSERT_FALSE(last->triggered);

	p->setCurrentFrame(20);
	p->tick();
	ASSERT_FALSE(first->triggered);
	ASSERT_TRUE(last->triggered);
}

TEST_F(PlayableTest, startFrame) {
	boost::shared_ptr<PlayableImpl> p(new PlayableImpl());
	p->setStartFrame(3);
	p->start();

	p->tick();
	ASSERT_FALSE(p->renderSingleCalled);

	p->tick();
	ASSERT_FALSE(p->renderSingleCalled);

	p->tick();
	ASSERT_FALSE(p->renderSingleCalled);

	p->tick();
	ASSERT_TRUE(p->renderSingleCalled);
	ASSERT_EQ(0, p->frameRendered);


	// the end frame should not be the duration, but the sum of the duration
	// and the start frame
	p->setCurrentFrame(10);
	p->tick();
	ASSERT_EQ(11, p->getCurrentFrame());


	// if we go to a frame greater than the duration, and loop is enabled it
	// should go back to frame zero and keep going from there
	p->setCurrentFrame(14);
	p->setLoop(true);
	p->tick();
	ASSERT_EQ(0, p->getCurrentFrame());

	p->tick();
	ASSERT_EQ(1, p->getCurrentFrame());

	// assures it stops if loop is disabled
	p->setLoop(false);
	p->setCurrentFrame(14);
	p->tick();
	ASSERT_FALSE(p->isPlaying());
}

File: fp_py/src/main/fruitpunch/native/elements/playable.py
from fruitpunch.native.elements.element import Element
import libfp_core
from fruitpunch.native.native_class import NativeClass

class Playable(Element):

    def __init__(self):
        NativeClass.__init__(self)
        
        self.store(libfp_core.Playable())
        
        self.listen({"render_single": self.on_render_single,
                     "calculate_duration": self.on_calculate_duration})

    # ---------------------------------------------------------------------
    # Native Methods
    # ---------------------------------------------------------------------

    # _____________________________________________________________________

    def start(self):
        return self.run_native("start")
    # _____________________________________________________________________

    def stop(self):
        return self.run_native("stop")

    # _____________________________________________________________________

    def get_current_frame(self):
        return self.run_native("get_current_frame")
    # _____________________________________________________________________

    def set_current_frame(self, currentFrame):
        return self.run_native("set_current_frame", currentFrame)
    # _____________________________________________________________________

    def is_playing(self):
        return self.run_native("is_playing")
    # _____________________________________________________________________

    def set_playing(self, isPlaying):
        return self.run_native("set_playing", isPlaying)
    # _____________________________________________________________________

    def is_loop(self):
        return self.run_native("is_loop")
    # _____________________________________________________________________

    def set_loop(self, loop):
        return self.run_native("set_loop", loop)
    # _____________________________________________________________________

    def get_start_frame(self):
        return self.run_native("get_start_frame")
    # _____________________________________________________________________

    def set_start_frame(self, startFrame):
        return self.run_native("set_start_frame", startFrame)
    # _____________________________________________________________________

    def get_end_frame(self):
        return self.run_native("get_end_frame")
    
    # _____________________________________________________________________
    
    def set_frame_duration(self, duration):
        return self.run_native("set_frame_duration", duration)
    
    # _____________________________________________________________________
    
    def get_frame_duration(self):
        return self.run_native("get_frame_duration")
    
    # _____________________________________________________________________
    
    def on_render_single(self, e):
        self.render_single(e.number)
    
    # _____________________________________________________________________
    
    def on_calculate_duration(self, e):
        e.number = self.calculate_duration()    
    
    # _____________________________________________________________________
    
    def render_single(self, frame_number):
        pass
    
    # _____________________________________________________________________
    
    def calculate_duration(self):
        pass

    # ---------------------------------------------------------------------
    # Properties
    # ---------------------------------------------------------------------

    current_frame = property(get_current_frame, set_current_frame)
    start_frame = property(get_start_frame, set_start_frame)
    end_frame = property(get_end_frame)
    loop = property(is_loop, set_loop)
    frame_duration = property(get_frame_duration, set_frame_duration)

NativeClass.class_map[libfp_core.Playable] = Playable